File: src/main/java/lazy/moheads/mixin/LivingEntityMixin.java
package lazy.moheads.mixin;

import lazy.moheads.event.LivingMobDrops;
import net.minecraft.world.damagesource.DamageSource;
import net.minecraft.world.entity.LivingEntity;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(LivingEntity.class)
public class LivingEntityMixin {

    @Inject(method = "dropAllDeathLoot", at = @At("TAIL"))
    public void dropAllDeathLoot(DamageSource source, CallbackInfo ci) {
        var result = LivingMobDrops.Companion.getEvent().invoker().addDrops((LivingEntity) (Object) this, source);
        if (source.getEntity() != null) {
            result.forEach(source.getEntity().level::addFreshEntity);
        }
    }
}

File: src/main/kotlin/lazy/moheads/mixin/LivingEntityMixin.kt
package lazy.moheads.mixin

import lazy.moheads.event.LivingMobDrops
import net.minecraft.world.damagesource.DamageSource
import net.minecraft.world.entity.LivingEntity
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

@Mixin(LivingEntity::class)
class LivingEntityMixin {

    @Inject(method = ["dropAllDeathLoot"], at = [At("TAIL")])
    fun dropAllDeathLoot(source: DamageSource, ci: CallbackInfo) {
        val result = LivingMobDrops.EVENT.invoker().addDrops((this as LivingEntity), source)
        source.entity?.run {
            result.forEach { this.level.addFreshEntity(it) }
        }
    }
}